accelerate-llvm-ptx/src/Data/Array/Accelerate/LLVM/PTX/CodeGen/Base.hs

```diff
@@ -325,6 +328,101 @@ atomicAdd_f t addr val =
 #endif
 
 
+-- Shuffles
+-- ---------
+
+-- CUDA has generic shuffles, but LLVM only provides i32 and f32 versions.
+-- To shuffle bigger types, we'll need to cast them into 32 bit segments,
+-- shuffle those, and then reassemble them. TODO
+shfl_up :: TypeR a
+        -> Operands a
+        -> Operands Word32
+        -> CodeGen PTX (Operands a)
+shfl_up = undefined
+
+shfl_down :: forall a. TypeR a
+          -> Operands a
+          -> Operands Word32
+          -> CodeGen PTX (Operands a)
+shfl_down typer a delta = case typer of
+  TupRunit -> return OP_Unit
+  TupRpair _ _ -> error "Acc.LLVM.PTX.CG.Base - I thought AoS->SoA conversion would prevent this"
+  TupRsingle sctyp -> case sctyp of
+    SingleScalarType (NumSingleType x) -> case x of
+      IntegralNumType TypeInt    -> cast_shfl_cast_i -- hope that Int isn't bigger than 32 bits! TODO
```

Comment:
Foreign.Storable.sizeOf (undefined::Int) == 8

Reply:
Right, oops, putting that one (and Word) in the 64-bit list

Reply:
Some of the Tegra boards are 32-bit, actually.